### PR TITLE
Add parsing for String/Long/Double Terms aggregations

### PR DESCRIPTION
**Pull request by tlrx**
This pull request adds parsing methods for Long/Double/String terms aggregations

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/ParsedTerms.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    @Override
+    public long getSumOfOtherDocCounts() {
+        return sumOtherDocCount;
+    }
+
+    private void setDocCountErrorUpperBound(long docCountErrorUpperBound) {
+        this.docCountErrorUpperBound = docCountErrorUpperBound;
+    }
+
+    private void setSumOtherDocCount(long sumOtherDocCount) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/ParsedTerms.java]: nit: maybe we can get rid of private setters and just set the field values where we need to set them?

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/ParsedStringTerms.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            return getKeyAsString();
+        }
+
+        @Override
+        public String getKeyAsString() {
+            String keyAsString = super.getKeyAsString();
+            if (keyAsString != null) {
+                return keyAsString;
+            }
+            return DocValueFormat.RAW.format((BytesRef) super.getKey());
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/ParsedStringTerms.java]: remove DocValueFormat dep here too?

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/ParsedLongTerms.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    public static class ParsedBucket extends ParsedTerms.ParsedBucket<Long> {
+
+        @Override
+        public String getKeyAsString() {
+            String keyAsString = super.getKeyAsString();
+            if (keyAsString != null) {
+                return keyAsString;
+            }
+            return DocValueFormat.RAW.format((Long) getKey());
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/ParsedLongTerms.java]: also remove DocValueFormat usage here too?

**Review by javanna — CHANGES_REQUESTED** [review]
left a few minor comments, LGTM otherwise

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/ParsedLongTerms.java (hunk 3 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/ParsedLongTerms.java]: > this casting is rather annoying. but I guess we can't do much about it as MultiBucketsAggregation.Bucket has getKey return an Object.

It also annoyed @cbuescher when he reviewed the parsing PR about histograms. I think I can remove the generic type and get rid of the casting at once, I'll do that.

**Comment by tlrx** [comment]
Thanks @javanna for your review. I updated the code according to your comments. Can you please have another look?

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ParsedDateHistogram.java [hunk 4]
```diff
@@ -56,25 +56,37 @@ public static ParsedDateHistogram fromXContent(XContentParser parser, String nam
...
 
         @Override
         public String getKeyAsString() {
             String keyAsString = super.getKeyAsString();
             if (keyAsString != null) {
                 return keyAsString;
-            } else {
-                return DocValueFormat.RAW.format((Long) super.getKey());
             }
+            if (key != null){
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ParsedDateHistogram.java]: missing whitespace after the parentheses :)

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/ParsedTerms.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        this.sumOtherDocCount = sumOtherDocCount;
+    }
+
+    @Override
+    public List<? extends Terms.Bucket> getBuckets() {
+        return buckets.stream().map(bucket -> (Terms.Bucket) bucket).collect(Collectors.toList());
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/ParsedTerms.java]: Yes, by adding a generic type to ParsedMultiBucketAggregation

**Review by javanna — APPROVED** [review]
looks great @tlrx thanks a lot!

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/ParsedTerms.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+                    } else if (CommonFields.KEY.getPreferredName().equals(currentFieldName)) {
+                        keyConsumer.accept(parser, bucket);
+                    } else if (CommonFields.DOC_COUNT.getPreferredName().equals(currentFieldName)) {
+                        bucket.setDocCount(parser.longValue());
+                    } else if (DOC_COUNT_ERROR_UPPER_BOUND_FIELD_NAME.getPreferredName().equals(currentFieldName)) {
+                        bucket.docCountError = parser.longValue();
+                        bucket.showDocCountError = true;
+                    }
+                } else if (token == XContentParser.Token.START_OBJECT) {
+                    aggregations.add(XContentParserUtils.parseTypedKeysObject(parser, Aggregation.TYPED_KEYS_DELIMITER, Aggregation.class));
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/ParsedTerms.java]: Side note: we talked about this earlier, if we encounter a "type#name" combination here where "type" isn't registered we will throw an exception, there is no way to ignore this at the moment. I don't know if we should add that option to `parseTypedKeysObject` and also to `parser#namedObject` 
This has nothing to do with this PR.

**Review by cbuescher — APPROVED** [review]
@tlrx thanks, after looking at this I think I start to understand the structure of how we parse multi bucket aggregations a lot better. Given the already complicated inheritance structure of the aggregations & buckets it looks great in using what we already have.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/ParsedTerms.java (hunk 6 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/ParsedTerms.java]: I think it's good thing that an exception is thrown. Handling uneregistered aggregation types is like handling unknown fields - for now we throw exceptions. If we want to manage those then I agree we will add to change the behavior of parseTypedKeysObject.

**Comment by tlrx** [comment]
Thanks @javanna @cbuescher !

**Pull request merged**